File: BackEnd/PHP/laravelshop/app/Http/Controllers/Api/V1/Auction/AuctionController.php
<?php

namespace App\Http\Controllers\Api\V1\Auction;

use App\Http\Controllers\Api\V1\ApiController;
use App\Modules\Auction\Requests\StoreAuctionBidRequest;
use App\Modules\Auction\Requests\StoreAuctionRequest;
use App\Modules\Auction\Requests\UpdateAuctionBidRequest;
use App\Modules\Auction\Services\AuctionService;
use Illuminate\Http\Request;

class AuctionController extends ApiController
{
    public function __construct(private readonly AuctionService $service)
    {
    }

    public function store(StoreAuctionRequest $request)
    {
        return $this->success($this->service->create($request->user(), $request->validated()), [], 201);
    }

    public function index(Request $request)
    {
        return $this->success($this->service->list($request->query('status'), $request->query('categoryId') ? (int) $request->query('categoryId') : null, (int) $request->query('page', 1), (int) $request->query('limit', 20)));
    }

    public function show(int $id)
    {
        return $this->success($this->service->detail($id));
    }

    public function storeBid(StoreAuctionBidRequest $request, int $id)
    {
        return $this->success($this->service->createBid($request->user(), $id, $request->validated()), [], 201);
    }

    public function selectBid(Request $request, int $id, int $bidId)
    {
        return $this->success($this->service->selectBid($request->user(), $id, $bidId));
    }

    public function destroy(Request $request, int $id)
    {
        return $this->success($this->service->cancel($request->user(), $id));
    }

    public function updateBid(UpdateAuctionBidRequest $request, int $id, int $bidId)
    {
        return $this->success($this->service->updateBid($request->user(), $id, $bidId, $request->validated()));
    }

    public function destroyBid(Request $request, int $id, int $bidId)
    {
        return $this->success($this->service->deleteBid($request->user(), $id, $bidId));
    }
}

File: BackEnd/Java/java-spring-maven-jpa-postgresql/src/main/java/com/pbshop/java/spring/maven/jpa/postgresql/auction/AuctionController.java
package com.pbshop.java.spring.maven.jpa.postgresql.auction;

import java.util.Map;

import jakarta.validation.Valid;

import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import com.pbshop.java.spring.maven.jpa.postgresql.auction.dto.AuctionDtos.CreateAuctionBidRequest;
import com.pbshop.java.spring.maven.jpa.postgresql.auction.dto.AuctionDtos.CreateAuctionRequest;
import com.pbshop.java.spring.maven.jpa.postgresql.auction.dto.AuctionDtos.UpdateAuctionBidRequest;
import com.pbshop.java.spring.maven.jpa.postgresql.auth.security.AuthenticatedUserPrincipal;
import com.pbshop.java.spring.maven.jpa.postgresql.common.api.ApiResponse;

@RestController
@RequestMapping("${pbshop.api.base-path:/api/v1}/auctions")
public class AuctionController {

    private final AuctionService auctionService;

    public AuctionController(AuctionService auctionService) {
        this.auctionService = auctionService;
    }

    @GetMapping
    public ApiResponse<Map<String, Object>> list(
            @RequestParam(required = false) String status,
            @RequestParam(required = false) Long categoryId,
            @RequestParam(defaultValue = "1") Integer page,
            @RequestParam(defaultValue = "20") Integer limit
    ) {
        return ApiResponse.success(auctionService.list(status, categoryId, page, limit));
    }

    @GetMapping("/{id}")
    public ApiResponse<Map<String, Object>> show(@PathVariable Long id) {
        return ApiResponse.success(auctionService.detail(id));
    }

    @PostMapping
    public ApiResponse<Map<String, Object>> create(
            @AuthenticationPrincipal AuthenticatedUserPrincipal principal,
            @Valid @RequestBody CreateAuctionRequest request
    ) {
        return ApiResponse.success(auctionService.create(principal, request));
    }

    @PostMapping("/{id}/bids")
    public ApiResponse<Map<String, Object>> createBid(
            @AuthenticationPrincipal AuthenticatedUserPrincipal principal,
            @PathVariable Long id,
            @Valid @RequestBody CreateAuctionBidRequest request
    ) {
        return ApiResponse.success(auctionService.createBid(principal, id, request));
    }

    @PatchMapping("/{id}/bids/{bidId}/select")
    public ApiResponse<Map<String, Object>> selectBid(
            @AuthenticationPrincipal AuthenticatedUserPrincipal principal,
            @PathVariable Long id,
            @PathVariable Long bidId
    ) {
        return ApiResponse.success(auctionService.selectBid(principal, id, bidId));
    }

    @DeleteMapping("/{id}")
    public ApiResponse<Map<String, Object>> delete(
            @AuthenticationPrincipal AuthenticatedUserPrincipal principal,
            @PathVariable Long id
    ) {
        return ApiResponse.success(auctionService.cancel(principal, id));
    }

    @PatchMapping("/{id}/bids/{bidId}")
    public ApiResponse<Map<String, Object>> updateBid(
            @AuthenticationPrincipal AuthenticatedUserPrincipal principal,
            @PathVariable Long id,
            @PathVariable Long bidId,
            @Valid @RequestBody UpdateAuctionBidRequest request
    ) {
        return ApiResponse.success(auctionService.updateBid(principal, id, bidId, request));
    }

    @DeleteMapping("/{id}/bids/{bidId}")
    public ApiResponse<Map<String, Object>> deleteBid(
            @AuthenticationPrincipal AuthenticatedUserPrincipal principal,
            @PathVariable Long id,
            @PathVariable Long bidId
    ) {
        return ApiResponse.success(auctionService.deleteBid(principal, id, bidId));
    }
}
